examples/ex9d.cpp

```diff
@@ -0,0 +1,510 @@
+      dc->SetCycle(0);
+      dc->SetTime(0.0);
+      dc->Save();
+   }
+
+   socketstream sout;
+   if (visualization)
+   {
+      char vishost[] = "localhost";
+      int  visport   = 19916;
+      sout.open(vishost, visport);
+      if (!sout)
+      {
+         cout << "Unable to connect to GLVis server at "
+              << vishost << ':' << visport << endl;
+         visualization = false;
+         cout << "GLVis visualization disabled.\n";
+      }
+      else
+      {
+         sout.precision(precision);
+         sout << "solution\n" << *mesh << u;
+         sout << "pause\n";
+         sout << flush;
+         cout << "GLVis visualization paused."
+              << " Press space (in the GLVis window) to resume it.\n";
+      }
+   }
+
+   // 8. Define the time-dependent evolution operator describing the ODE
+   //    right-hand side, and perform time-integration (looping over the time
+   //    iterations, ti, with a time-step dt).
+   FE_Evolution adv(M, K, B);
+
+   double t = 0.0;
+   adv.SetTime(t);
+   ode_solver->Init(adv);
+
+   tic_toc.Clear();
+   tic_toc.Start();
+
+   bool done = false;
+   for (int ti = 0; !done; )
+   {
+      double dt_real = min(dt, t_final - t);
+      ode_solver->Step(u, t, dt_real);
+      ti++;
+
+      done = (t >= t_final - 1e-8*dt);
+
+      if (done || ti % vis_steps == 0)
+      {
+         cout << "time step: " << ti << ", time: " << t << endl;
+
+         if (visualization)
+         {
+            sout << "solution\n" << *mesh << u << flush;
+         }
+
+         if (visit)
+         {
+            dc->SetCycle(ti);
+            dc->SetTime(t);
+            dc->Save();
+         }
+      }
+   }
+
+   tic_toc.Stop();
+   cout << " done, " << tic_toc.RealTime() << "s." << endl;
+
+   // 9. Save the final solution. This output can be viewed later using GLVis:
+   //    "glvis -m ex9.mesh -g ex9-final.gf".
+   {
+      ofstream osol("ex9-final.gf");
+      osol.precision(precision);
+      u.Save(osol);
+   }
+
+   // 10. Free the used memory.
+   delete ode_solver;
+   delete dc;
+
+   return 0;
+}
+
+
+// Implementation of class FE_Evolution
+// FE_Evolution::FE_Evolution(SparseMatrix &_M, SparseMatrix &_K, const Vector &_b)
+//    : TimeDependentOperator(_M.Size()), M(_M), K(_K), b(_b), z(_M.Size())
+// {
+//    M_solver.SetPreconditioner(M_prec);
+//    M_solver.SetOperator(M);
+
+//    M_solver.iterative_mode = false;
+//    M_solver.SetRelTol(1e-9);
+//    M_solver.SetAbsTol(0.0);
+//    M_solver.SetMaxIter(100);
+//    M_solver.SetPrintLevel(0);
+// }
+FE_Evolution::FE_Evolution(OperatorHandle &_M, OperatorHandle &_K,
+                           const Vector &_b)
+   : TimeDependentOperator(*_K.Ptr()), M(_M), K(_K), b(_b), z(_b,false)
+{
+}
+
+void FE_Evolution::Mult(const Vector &x, Vector &y) const
+{
+   // y = M^{-1} (K x + b)
+   K.Ptr()->Mult(x, z);
+   // z += b;
+   z.Axpby(1.0, z, 1.0, b);
+   // M.Mult(z, y);
+   CG(*M.Ptr(), z, y, 0, 1000, 1e-12, 0.0);
```

Comment:
There is some setup cost to calling the function `CG` -- it will allocate a new `CGSolver` every time. That is why the original example has a `CGSolver` (called `M_solver`) as part of the `FE_Evolution` class which is initialized once. Not sure if this is important for CPU but for GPU the memory allocations (performed when a `CGSolver` is initialized) will definitely slow things down.